### ISSUE TITLE
Dispatcher TLS to bypass loads of dispatch key from tensor arguments

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#54614]
@ailzhang recently wrote in https://github.com/pytorch/rfcs/pull/17 about how, after the introduction of InplaceOrView key, we are tantalizingly close to removing the Autograd key from tensors with `requires_grad=False`. However, the main problem is this:

> If requires_grad=False skips VariableType kernel, functional ops won't be able to go through AutoDispatchBelowInplaceOrView guard which suppresses both autograd and InplaceOrView keys in TLS excluded. Not suppressing InplaceOrView key means unnecessary calls to as_view/increment_version if any view/inplace ops are used in the kernel implementation which adds a lot of overhead. To avoid overhead, instead of fallthrough kerenl being backend fallback, we'll want to use a real kernel that suppresses InplaceOrView key. But compared to the current implementation which only adds an extra dispatch for view/inplace ops, it forces all functional ops to have an extra dispatch as well. That's why it's blocked.

I was thinking about ways around this problem and realized that an old proposal (that I am on the record as not liking--sorry, I don't remember who first proposed it, perhaps it was @bhosmer) could help solve this problem.

The proposal was that once you get to a "backend" kernel, you probably only want to ever stay dispatching in that kernel (for example, you don't want to skip back to InplaceOrView). So there should be some TLS flag that, when set, bypasses argument checking; instead, we always dispatch to the dispatch key stored in the TLS. This flag would be set on entry to backend kernels (via the code generated wrappers that also take care of things like device guard). This now solves Ailing's problem as we now can bypass those view/increment_version operations.

One thing that I'm still a little uncomfortable about is that we have to be *absolutely absolutely* certain that we're never going to want to pop out somewhere else inside a backend implementation; it would be pretty awkward if occasionally this is something we needed to do. And of course, the other downside is we're adding an extra unconditional branch on the dispatcher path (I'm not even sure the branch is easy to predict).

**Comment by ailzhang** [comment]
Yup this might indeed make it possible, plus that we will likely get some perf win from "stop going back to upper level of disaptcher - TLS write overhead".  :D

**Comment by ngimel** [comment]
With current profiler implementation, it disallows us to peek into internal implementation of the op, which was also one of the arguments against this proposal. But perf wins might be too tempting :-)

**Comment by ailzhang** [comment]
@ngimel If that's a concern, maybe we can provide an option in profiler switch between two modes? (it'll be mostly toggling TLS I think)

**Comment by ngimel** [comment]
Yeah, that's a good option.

**Comment by ezyang** [comment]
> With current profiler implementation, it disallows us to peek into internal implementation of the op, which was also one of the arguments against this proposal. But perf wins might be too tempting :-)

Also, this TLS setter could happen at the beginning of the kernel, rather than in the dispatcher itself
